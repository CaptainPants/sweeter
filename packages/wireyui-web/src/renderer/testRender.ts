import { WebRenderer } from './WebRenderer.js';

export interface TestRenderResult {
    readonly nodes: NodeListOf<ChildNode>;
    readonly getHTML: () => string;
    readonly dispose: () => void;
}

/**
 * Convenience function for rendering isolated JSX chunks for unit tests
 * @param callback
 * @returns
 */
export function testRender(callback: () => JSX.Element): TestRenderResult {
    const renderer = new WebRenderer();

    const root = document.createElement('div');

    const dispose = renderer.start(root, callback);

    return {
        nodes: root.childNodes,
        getHTML: () => root.innerHTML,
        dispose,
    };
}
