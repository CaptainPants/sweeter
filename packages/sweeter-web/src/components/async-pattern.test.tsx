/* @jsxImportSource .. */

import {
    $async,
    $val,
    type Component,
    Suspense,
} from '@captainpants/sweeter-core';
import { testRender } from '../test/testRender.js';

interface Component1Props {
    dataLoadComplete: AbortSignal;
    rendered: AbortController;
}

const Component1: Component<Component1Props> = (props, init) => {
    const rendered = $async(
        (abort) => {
            return new Promise<string>((resolve) => {
                props.dataLoadComplete.addEventListener('abort', () => {
                    resolve('RESOLVED!');
                });
            });
        },
        (asyncInitializerResult) => {
            return <div>RESULT: {$val(asyncInitializerResult)}</div>;
        },
    );

    props.rendered.abort();

    return rendered;
};

it('example 1', async () => {
    const release = new AbortController();
    const rendered = new AbortController();

    const res = testRender(() => (
        <Suspense fallback={() => <div>LOADING...</div>}>
            {() => (
                <Component1
                    dataLoadComplete={release.signal}
                    rendered={rendered}
                />
            )}
        </Suspense>
    ));

    expect(res.getHTML()).toMatchSnapshot();

    release.abort();

    await new Promise((resolve) => {
        rendered.signal.addEventListener('abort', () => {
            resolve(void 0);
        });
    });

    expect(res.getHTML()).toMatchSnapshot();

    res.dispose();
});
