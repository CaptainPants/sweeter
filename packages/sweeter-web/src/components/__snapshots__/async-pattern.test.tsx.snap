// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`example 1 1`] = `"<div>LOADING...</div><div style=\\"display: none;\\"><!--Hooks(Async): trackSignals--></div>"`;

exports[`example 1 2`] = `"<div>RESULT: RESOLVED!</div><!--Hooks(Async): trackSignals-->"`;
